### DISCUSSION TITLE
Can't run mike versioning

### DISCUSSION THREAD

**Discussion by berkantsahn** [squidfunk/mkdocs-material#5356] [Q&A] [answered]
Hi. I am trying to build a sample with mkdocs material and i want to add versioning to my project. I have two different language options in the project and now trying to add versioning with mike.

I installed mike with pip and try to run
```
mike deploy --push --update-aliases 0.1 latest
```

but it didn't work and it gave an error like this; "mike : The term 'mike' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was includ
ed, verify that the path is correct and try again."

There is my project repo link: https://github.com/berkantsahn/mkdocs-sample

Thanks in advance for your help.

**Comment by kamilkrzyskow** [comment]
The error indicates a lack of installed `mike`.

https://pypi.org/project/mike/

If it's installed and somehow didn't register as a standalone application, then you could try running it via python:

```
python -m mike ....
```

Though I never used it neither with normal docs nor with multilingual docs.

**Reply by berkantsahn** [reply]
When I want to run 
```
python -m mike ....
```
it gives an error like this; No module named mike.__main__; 'mike' is a package and cannot be directly executed

But i already installed mike. Idk why is it not working.

**Reply by kamilkrzyskow** [reply]
OK, seems like the `-m` approach won't work here, as mike doesn't have any module, which processes the args directly. Maybe there is a way to call the `mike.driver.main()` function somehow from the Terminal, but I don't know it.
If `mike` is installed, but there is an error like `mike is not recognized as cmdlet or script` then your directory with installed scripts isn't in the system PATH. 
You can run:
```shell
python
>>> import sys
>>> print(*sys.path, sep="\n")
```
In the list of paths find the one containing a `Lib` directory, e.g. `C:\SomeFiles\Python311\Lib`
Then make sure that the `C:\SomeFiles\Python311\Scripts` directory next to it is added to your system PATH.
The `mike` executable should be located in the `Scripts` directory, and then it being in the PATH would allow to access it anywhere in the system via the `mike` command and there should be no more errors about it being unrecognized.

---

Overall a good practice is to create a Virtual Environment and activate it, as it temporarily adds all required paths to the PATH.
There is even a guide here: 
https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/#environment

**Reply by berkantsahn** [reply]
I already added C:\SomeFiles\Python311\Scripts to my system path when i installed the python. So when I looked in the Lib and Scripts folder, I couldn't find anything about mike, maybe that's why the problem happens and then when I want to reinstall mike, I observe that it is already installed.

**Reply by kamilkrzyskow** [reply, marked as answer]
Did you set up the Virtual Environment? You can also run this command to reinstall `mike`:
```shell
pip install mike --force-reinstall --verbose
```
My output looks like this:
```shell
Using pip 23.0.1 from C:\MyFiles\Python311\Lib\site-packages\pip (python 3.11)
Collecting mike
  Using cached mike-1.1.2-py3-none-any.whl (26 kB)
...
Installing collected packages: verspec, watchdog, six, pyyaml, packaging, mergedeep, MarkupSafe, markdown, colorama, pyyaml-env-tag, python-dateutil, jinja2, click, ghp-import, mkdocs, mike
...
  Attempting uninstall: mkdocs
    Found existing installation: mkdocs 1.4.2
    Uninstalling mkdocs-1.4.2:
      Removing file or directory c:\myfiles\python311\lib\site-packages\mkdocs-1.4.2.dist-info\
      Removing file or directory c:\myfiles\python311\lib\site-packages\mkdocs\
      Removing file or directory c:\myfiles\python311\scripts\mkdocs.exe
      Successfully uninstalled mkdocs-1.4.2
  Attempting uninstall: mike
    Found existing installation: mike 1.1.2
    Uninstalling mike-1.1.2:
      Removing file or directory c:\myfiles\python311\lib\site-packages\mike-1.1.2.dist-info\
      Removing file or directory c:\myfiles\python311\lib\site-packages\mike\
      Removing file or directory c:\myfiles\python311\scripts\mike.exe
      Successfully uninstalled mike-1.1.2
Successfully installed MarkupSafe-2.1.2 click-8.1.3 colorama-0.4.6 ghp-import-2.1.0 jinja2-3.1.2 markdown-3.3.7 mergedeep-1.3.4 mike-1.1.2 mkdocs-1.4.2 packaging-23.0 python-dateutil-2.8.2 pyyaml-6.0 pyyaml-env-tag-0.1 six-1.16.0 verspec-0.1.0 watchdog-3.0.0
```
As seen in the log `scripts\mike.exe` got deleted and this file should be run via the `mike` command, because the `scripts` path is added to the PATH 
![image](https://user-images.githubusercontent.com/34622465/231454142-cf610733-d929-4e5e-9d46-d4c0317535df.png)

**Reply by kamilkrzyskow** [reply]
Also this is the answer thread to the error in the OP don't forgot to mark it as the answer ✌️

**Comment by berkantsahn** [comment]
While reinstalling `mike` ,  I ran with following commands
```
pip install mike --force-reinstall --verbose
```

So this is my output when i ran the command;
```
Using pip 23.0.1 from C:\Python311\Lib\site-packages\pip (python 3.11)
Collecting mike
  Using cached mike-1.1.2-py3-none-any.whl (26 kB)
Collecting jinja2
  Using cached Jinja2-3.1.2-py3-none-any.whl (133 kB)
Collecting mkdocs>=1.0
  Using cached mkdocs-1.4.2-py3-none-any.whl (3.7 MB)
Collecting pyyaml>=5.1
  Using cached PyYAML-6.0-cp311-cp311-win_amd64.whl (143 kB)
Collecting verspec
  Using cached verspec-0.1.0-py3-none-any.whl (19 kB)
Collecting click>=7.0
  Using cached click-8.1.3-py3-none-any.whl (96 kB)
Collecting colorama>=0.4
  Using cached colorama-0.4.6-py2.py3-none-any.whl (25 kB)
Collecting ghp-import>=1.0
  Using cached ghp_import-2.1.0-py3-none-any.whl (11 kB)
Collecting markdown<3.4,>=3.2.1
  Using cached Markdown-3.3.7-py3-none-any.whl (97 kB)
Collecting mergedeep>=1.3.4
  Using cached mergedeep-1.3.4-py3-none-any.whl (6.4 kB)
Collecting packaging>=20.5
  Using cached packaging-23.0-py3-none-any.whl (42 kB)
Collecting pyyaml-env-tag>=0.1
  Using cached pyyaml_env_tag-0.1-py3-none-any.whl (3.9 kB)
  Link requires a different Python (3.11.0 not in: '>=2.7, <3.6'): https://files.pythonhosted.org/packages/8e/68/261843fe0fed52c7f4364325bde666c858c582547c6a32e90ce9c184a5d7/watchdog-0.10.6.tar.gz (from https://pypi.org/simple/watchdog/) (requires-python:>=2.7, <3.6)
  Link requires a different Python (3.11.0 not in: '>=2.7, <3.6'): https://files.pythonhosted.org/packages/9d/f8/99b803268b9cc809df0934c54266777db3ab7698ff6c82021a00e4fdfc03/watchdog-0.10.7.tar.gz (from https://pypi.org/simple/watchdog/) (requires-python:>=2.7, <3.6)
Collecting watchdog>=2.0
  Using cached watchdog-3.0.0-py3-none-win_amd64.whl (82 kB)
Collecting MarkupSafe>=2.0
  Using cached MarkupSafe-2.1.2-cp311-cp311-win_amd64.whl (16 kB)
Collecting python-dateutil>=2.8.1
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Collecting six>=1.5
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Installing collected packages: verspec, watchdog, six, pyyaml, packaging, mergedeep, MarkupSafe, markdown, colorama, pyyaml-env-tag, python-dateutil, jinja2, click, ghp-import, mkdocs, mike
  Attempting uninstall: verspec
    Found existing installation: verspec 0.1.0
    Uninstalling verspec-0.1.0:
      Removing file or directory c:\python311\lib\site-packages\verspec-0.1.0.dist-info\
      Removing file or directory c:\python311\lib\site-packages\verspec\
      Successfully uninstalled verspec-0.1.0
  Attempting uninstall: watchdog
    Found existing installation: watchdog 3.0.0
    Uninstalling watchdog-3.0.0:
      Removing file or directory c:\python311\lib\site-packages\watchdog-3.0.0.dist-info\
      Removing file or directory c:\python311\lib\site-packages\watchdog\
      Successfully uninstalled watchdog-3.0.0
  Attempting uninstall: six
    Found existing installation: six 1.16.0
    Uninstalling six-1.16.0:
      Removing file or directory c:\python311\lib\site-packages\__pycache__\six.cpython-311.pyc
      Removing file or directory c:\python311\lib\site-packages\six-1.16.0.dist-info\
      Removing file or directory c:\python311\lib\site-packages\six.py
      Successfully uninstalled six-1.16.0
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 6.0
    Uninstalling PyYAML-6.0:
      Removing file or directory c:\python311\lib\site-packages\_yaml\
      Removing file or directory c:\python311\lib\site-packages\pyyaml-6.0.dist-info\
      Removing file or directory c:\python311\lib\site-packages\yaml\
      Successfully uninstalled PyYAML-6.0
  Attempting uninstall: packaging
    Found existing installation: packaging 23.0
    Uninstalling packaging-23.0:
      Removing file or directory c:\python311\lib\site-packages\packaging-23.0.dist-info\
      Removing file or directory c:\python311\lib\site-packages\packaging\
      Successfully uninstalled packaging-23.0
  Attempting uninstall: mergedeep
    Found existing installation: mergedeep 1.3.4
    Uninstalling mergedeep-1.3.4:
      Removing file or directory c:\python311\lib\site-packages\mergedeep-1.3.4.dist-info\
      Removing file or directory c:\python311\lib\site-packages\mergedeep\
      Successfully uninstalled mergedeep-1.3.4
  Attempting uninstall: MarkupSafe
    Found existing installation: MarkupSafe 2.1.2
    Uninstalling MarkupSafe-2.1.2:
      Removing file or directory c:\python311\lib\site-packages\markupsafe-2.1.2.dist-info\
      Removing file or directory c:\python311\lib\site-packages\markupsafe\
      Successfully uninstalled MarkupSafe-2.1.2
  Attempting uninstall: markdown
    Found existing installation: Markdown 3.3.7
    Uninstalling Markdown-3.3.7:
      Removing file or directory c:\python311\lib\site-packages\markdown-3.3.7.dist-info\
      Removing file or directory c:\python311\lib\site-packages\markdown\
      Successfully uninstalled Markdown-3.3.7
  Attempting uninstall: colorama
    Found existing installation: colorama 0.4.6
    Uninstalling colorama-0.4.6:
      Removing file or directory c:\python311\lib\site-packages\colorama-0.4.6.dist-info\
      Removing file or directory c:\python311\lib\site-packages\colorama\
      Successfully uninstalled colorama-0.4.6
  Attempting uninstall: pyyaml-env-tag
    Found existing installation: pyyaml_env_tag 0.1
    Uninstalling pyyaml_env_tag-0.1:
      Removing file or directory c:\python311\lib\site-packages\__pycache__\yaml_env_tag.cpython-311.pyc
      Removing file or directory c:\python311\lib\site-packages\pyyaml_env_tag-0.1.dist-info\
      Removing file or directory c:\python311\lib\site-packages\yaml_env_tag.py
      Successfully uninstalled pyyaml_env_tag-0.1
  Attempting uninstall: python-dateutil
    Found existing installation: python-dateutil 2.8.2
    Uninstalling python-dateutil-2.8.2:
      Removing file or directory c:\python311\lib\site-packages\dateutil\
      Removing file or directory c:\python311\lib\site-packages\python_dateutil-2.8.2.dist-info\
      Successfully uninstalled python-dateutil-2.8.2
  Attempting uninstall: jinja2
    Found existing installation: Jinja2 3.1.2
    Uninstalling Jinja2-3.1.2:
      Removing file or directory c:\python311\lib\site-packages\jinja2-3.1.2.dist-info\
      Removing file or directory c:\python311\lib\site-packages\jinja2\
      Successfully uninstalled Jinja2-3.1.2
  Attempting uninstall: click
    Found existing installation: click 8.1.3
    Uninstalling click-8.1.3:
      Removing file or directory c:\python311\lib\site-packages\click-8.1.3.dist-info\
      Removing file or directory c:\python311\lib\site-packages\click\
      Successfully uninstalled click-8.1.3
  Attempting uninstall: ghp-import
    Found existing installation: ghp-import 2.1.0
    Uninstalling ghp-import-2.1.0:
      Removing file or directory c:\python311\lib\site-packages\__pycache__\ghp_import.cpython-311.pyc
      Removing file or directory c:\python311\lib\site-packages\ghp_import-2.1.0.dist-info\
      Removing file or directory c:\python311\lib\site-packages\ghp_import.py
      Removing file or directory c:\python311\scripts\ghp-import.exe
      Successfully uninstalled ghp-import-2.1.0
  Attempting uninstall: mkdocs
    Found existing installation: mkdocs 1.4.2
    Uninstalling mkdocs-1.4.2:
      Removing file or directory c:\python311\lib\site-packages\mkdocs-1.4.2.dist-info\
      Removing file or directory c:\python311\lib\site-packages\mkdocs\
      Removing file or directory c:\python311\scripts\mkdocs.exe
      Successfully uninstalled mkdocs-1.4.2
  Attempting uninstall: mike
    Found existing installation: mike 1.1.2
    Uninstalling mike-1.1.2:
      Removing file or directory c:\python311\lib\site-packages\mike-1.1.2.dist-info\
      Removing file or directory c:\python311\lib\site-packages\mike\
      Successfully uninstalled mike-1.1.2
Successfully installed MarkupSafe-2.1.2 click-8.1.3 colorama-0.4.6 ghp-import-2.1.0 jinja2-3.1.2 markdown-3.3.7 mergedeep-1.3.4 mike-1.1.2 mkdocs-1.4.2 packaging-23.0 python-dateutil-2.8.2 pyyaml-6.0 pyyaml-env-tag-0.1 six-1.16.0 verspec-0.1.0 watchdog-3.0.0
```
and this is my environment variables window;

![env-variables](https://user-images.githubusercontent.com/37020037/231471366-85141959-1bbb-4112-92bc-8c79af1328d2.png)


I want to run versioning with github pages. I still couldn't find where I made the mistake.

**Reply by berkantsahn** [reply]
I made some progress on versioning now i realized when i reinstalled `mike` my command worked:
```
mike deploy --push --update-aliases 1.0 latest
```

But i still have an issues about versioning. when i run the `mike` deploy command.  It gives an error like this:
`
error: failed to push branch gh-pages to origin: "To https://github.com/berkantsahn/mkdocs-sample.git
 ! [rejected]        gh-pages -> gh-pages (non-fast-forward)
error: failed to push some refs to 'https://github.com/berkantsahn/mkdocs-sample.git'
hint: Updates were rejected because a pushed branch tip is behind its remote
hint: counterpart. Check out this branch and integrate the remote changes
hint: (e.g. 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details."
`

I can see the version folders are created but I couldn't get it to work.

**Reply by kamilkrzyskow** [reply]
Awesome! New error means progress 😎

Though I am curious how did you install `mike` without the executable 🤔Maybe you did some kind of system restoration after installing `mike`, which deleted the `.exe`.

The current error `non-fast forward` means that the remote `gh-pages` branch already uploaded to GitHub is not a direct history of your current setup and your local version isn't synced with it.

It's likely because you have the non-mike version already there.
One solution would be to delete the `gh-pages` branch and run the command again.

Another solution maybe is running `mike` with some kind of `--force` flag option to overwrite the remote branch.

---

Though as a side note, I still didn't try `mike` myself, I'm not sure what is the workflow of using mike properly :v: etc. how to assure sync with remote branches.

**Comment by berkantsahn** [comment]
Actually i installed `mike` with another command. I don't understand why I'm facing something like this but thanks it worked.

Coming to the current situation, yes, I have a branch called gh-pages on github.
When i try to sync with `vscode` i get an error:
![error1](https://user-images.githubusercontent.com/37020037/231483572-8d8e15d3-42eb-48d1-b86c-31a8c4730b67.png)

and when I try to pull, I get an error like this
![error2](https://user-images.githubusercontent.com/37020037/231485493-9075d0a5-def4-4d77-bd4b-4bc8a92120c1.png)


but I can observe that version folders are created on local.

**Reply by kamilkrzyskow** [reply]
I guess that the second error about `no tracking information` is, because you forgot to fetch? 
Try `git fetch origin` and then `git pull origin`. You can also run `git checkout -b --track origin/gh-pages` to make sure you have a local backup copy of the remote branch and delete the remote branch altogether via the GitHub interface https://github.com/berkantsahn/mkdocs-sample/branches

**Reply by berkantsahn** [reply]
I managed to sync branches and my version folders have been successfully uploaded and published on github. I got it to display after the folder is created on gh-pages and I don't have a problem switching from the first version to the next, but the version bar disappears when downgrading from the last version to first but it's ok for now i can handle it but I'm having problems with routing since I use two different languages. When I change between versions, there is no proper redirection. Do you have an opinion on this subject?

**Reply by kamilkrzyskow** [reply]
The current version of the branch has no versioning https://github.com/berkantsahn/mkdocs-sample/tree/gh-pages therefore I can't really say for sure. `mike` creates a `versions.json` file which is then used by the material theme to setup the version switch toggle
https://github.com/squidfunk/mkdocs-material/blob/2b2764a37196651f3057f77a725ac112b863624b/src/assets/javascripts/integrations/version/index.ts#L76

I assume that the link url creation doesn't take into account your `/lang_code/` path part from the multi-language setup. I do see that your `mkdocs.yml` files do lack the [`site_url`](https://www.mkdocs.org/user-guide/configuration/#site_url) option, but then again I doubt adding it will solve it here.

Change the title of this Q&A thread to `Can't run mike versioning`, mark my initial answer as the answer, as the issue was with an incorrect `mike` installation, and create another Q&A thread with a specific name `multi-language docs with mike versioning don't work`. ✌️ 
I never used `mike`, so I don't know where the issue could be, I only saw that the error in the OP was unrelated to mike, just generic installation issues, therefore I stepped in.

**Reply by squidfunk** [reply]
@kamilkrzyskow thanks for taking the time investigating and I agree – it's really unrelated to setup and was likely more a general setup issue. I changed the title accordingly and selected your answer.

---

On a general note I want to thank you for taking the time to provide high quality answers to discussions like this in the last weeks. I'm sponsoring several developers that maintain upstream (and downstream) dependencies of Material for MkDocs like Python Markdown Extensions, MkDocstrings and several plugin authors. I would be more than happy to upstream some of the funding Material for MkDocs receives via sponsorships to compensate for your time if you decide to set up a GitHub Sponsors account.

Your work and time is very appreciated!

**Reply by kamilkrzyskow** [reply]
@squidfunk Thanks for managing all of this, and for the proposal on setting up the sponsorship. I have no plans to do that right now, and I'll very likely be less active since ~June, maybe temporarily (weeks-months) or I'll get busier with life and won't be able to hang around like I'm doing now. 😞

Maybe someday I might set up sponsorship, when I'll get to creating the plugins for dynamic theme modification and git redirects, and then get 5-10 stars after seeing some popularity. Additionally I'm not that selfless, as I typically help out only when I expect the solution to be "easy", as in using knowledge I already have, or learning along the way knowing I might use the learned knowledge in the future in my own projects, so I don't think I need any compensation. 😸

For example here, forking Berkant's repository and debugging `mike` with the multilingual docs could be "difficult" and I'm also not really going to use the solution myself, so I'm rather reluctant helping out more. 😓

For example when I first wrote to you about the Co-Authoring in the Polish translation thread, I considered the issue "easy", while using automated IDE myself, and it turned out to not be so straight-forward, as I imagined... 😅

For example in [this discussion](https://github.com/mkdocs/mkdocs/discussions/3153) thread I went on a deep dive solution, because I wanted to learn about Jinja2 filter functions. However, I was a little hard on the OP, saying that his problem is kind of self-created in his [issue thread](https://github.com/mkdocs/mkdocs/issues/3140). I didn't get any feedback though, maybe because my initial answer came of as accusatory or passive-aggressive, or he didn't need to use it 🤷 I still got to use a custom Jinja2 function. 😸 

Why did I write all of this? idk 🤷, some kind of impostor syndrome, or just wanting to keep it real ✌️

**Reply by berkantsahn** [reply]
@kamilkrzyskow Thank you for all your help i am really appreciate. First of all, `mike` versioning problem was solved and I'm happy about it. As you said, I will create a different topic about multilingualism and continue from there.
Thanks again.
